pkg/reconciler/scheduling/namespace/namespace_reconcile_scheduling.go

```diff
@@ -0,0 +1,266 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package namespace
+
+import (
+	"context"
+	"encoding/json"
+	"math/rand"
+	"strings"
+	"time"
+
+	"github.com/kcp-dev/logicalcluster"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/klog/v2"
+	"k8s.io/kube-openapi/pkg/util/sets"
+
+	schedulingv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/scheduling/v1alpha1"
+	workloadv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/workload/v1alpha1"
+	locationreconciler "github.com/kcp-dev/kcp/pkg/reconciler/scheduling/location"
+)
+
+const removingGracePeriod = 10 * time.Second
+
+// placementSchedulingReconciler schedules a workload for this ns. It checks the current placement annotation on the ns,
+// and find all valid workloadclusters.
+type placementSchedulingReconciler struct {
+	listWorkloadCluster func(clusterName logicalcluster.Name) ([]*workloadv1alpha1.WorkloadCluster, error)
+
+	getPlacement func(clusterName logicalcluster.Name, name string) (*schedulingv1alpha1.Placement, error)
+
+	getLocation func(clusterName logicalcluster.Name, name string) (*schedulingv1alpha1.Location, error)
+
+	patchNamespace func(ctx context.Context, clusterName logicalcluster.Name, name string, pt types.PatchType, data []byte, opts metav1.PatchOptions, subresources ...string) (*corev1.Namespace, error)
+
+	enqueueAfter func(logicalcluster.Name, *corev1.Namespace, time.Duration)
+}
+
+func (r *placementSchedulingReconciler) reconcile(ctx context.Context, ns *corev1.Namespace) (reconcileStatus, *corev1.Namespace, error) {
+	clusterName := logicalcluster.From(ns)
+
+	var disableScheduling bool
+
+	if value, found := ns.Labels[workloadv1alpha1.SchedulingDisabledLabel]; found && value == "true" {
```

Comment:
we have to decide what to do with this label. Do we really need it now that we need a `Placement` object in the first place to start placement? I think the reason at the time for the label was to stop the resource scheduler to blindly schedule resources.

Reply:
we do not, we can use placement to control whether we want to disable scheduling for a ns.